** Build
[[https://travis-ci.org/BioInf-Wuerzburg/AliTV/][https://travis-ci.org/BioInf-Wuerzburg/AliTV.svg?branch=master]]
[[https://coveralls.io/github/BioInf-Wuerzburg/AliTV?branch=master][https://coveralls.io/repos/github/BioInf-Wuerzburg/AliTV/badge.svg?branch=master]]

** Homepage
[[http://bioinf-wuerzburg.github.io/AliTV][AliTV]]
** Installation

#+BEGIN_EXAMPLE
  git clone --recursive https://github.com/BioInf-Wuerzburg/AliTV
#+END_EXAMPLE
   
** Dependencies

- lastz or MUMmer
- Circos
- optional - blastn (for IR annotation)

** Usage
To get usage information
#+BEGIN_EXAMPLE
  AliTV --help
#+END_EXAMPLE

Example usage
#+BEGIN_EXAMPLE
  AliTV --query query.fa --reference reference.fa
#+END_EXAMPLE

** Output
By default, AliTV generates ten files in the output folder:

| <prefix>.txt             | The output of the whole genome alignment program in tabular format      |
| <prefix>.svg             | The circular map of the wga created by Circos in svg format             |
| <prefix>.png             | The circular map of the wga created by Circos in png format             |
| <prefix>.circos.conf     | The generated circos config file                                        |
| <prefix>.karyo           | The karyotype file for Circos                                           |
| <prefix>.link            | The link file for Circos                                                |
| <prefix>.high            | The highlight file for Circos                                           |
| <prefix>.wanted_ids      | The new IDs of sequences with illegal characters replaced by underscore |
| <prefix>_query.fasta     | A copy of the query fasta with adjusted IDs                             |
| <prefix>_reference.fasta | A copy of the reference fasta with adjusted IDs                         |

A demo of the new interactive HTML/JavaScript output can be found [[http://bioinf-wuerzburg.github.io/AliTV/d3/AliTV.html][here]].

** Input
Both the query and the reference file should be in fasta format.

** Citing AliTV

We have not yet submitted a manuscript.
Please use this repository to cite AliTV if you use it in your project.
[[https://zenodo.org/badge/latestdoi/12731/BioInf-Wuerzburg/AliTV][https://zenodo.org/badge/12731/BioInf-Wuerzburg/AliTV.svg]]

Please also consider citing:
 - lastz or MUMmer
 - Circos
 - BLAST (for IR annotation)

For a list of included libraries together with their respective licenses see the [[file:LICENSES/][LICENSES]] folder.
** Contact
If you have any questions or encounter problems or potential bugs, don't
hesitate to contact us. Either report [[https://github.com/BioInf-Wuerzburg/AliTV/issues][issues]] on github or write an email to:

- Markus Ankenbrand - markus.ankenbrand@uni-wuerzburg.de






#+TITLE: AliTV manual
#+AUTHOR: Markus Ankenbrand
#+EMAIL: markus.ankenbrand@uni-wuerzburg.de
#+LANGUAGE: en
#+OPTIONS: ^:nil date:nil H:2
#+LaTeX_CLASS: scrartcl
#+LaTeX_CLASS_OPTIONS: [a4paper,12pt,headings=small]
#+LaTeX_HEADER: \setlength{\parindent}{0pt}
#+LaTeX_HEADER: \setlength{\parskip}{1.5ex}
#+LATEX_HEADER: \renewcommand{\familydefault}{\sfdefault}
** Changelog
*** 0.3.5 <2015-09-24 Do>
 - Add version number to html page
 - Auto adjust tree representation when reordering genomes
 - Warn if tree is not concordant with genome order
 - Legend for biological feature types
 - Color scale for link identity
*** 0.3.4 <2015-09-16 Mi>
 - Add custom features via the interface
*** 0.3.3 <2015-09-11 Fr>
 - Bug fix in template
*** 0.3.2 <2015-09-11 Fr>
 - rotation of circular sequences
 - dynamic (biological) feature menus
 - code cleanup
*** 0.3.1 <2015-08-13 Do>
 - Strand specific features
 - Reset for genome zooming
 - Libraries for contextMenu are update with generateJSONFiles.pl
 - Fix #52: allow hiding of features with non-unique names
 - Add new alerts and error messages
*** 0.3 <2015-07-31 Fr>
 - Show selected genome regions (independent)
 - Horizontal scroll bar
 - Add complete tab structure for the interface
 - Add context menus for chromosomes, features and links
 - Allow reordering of genomes and chromosomes by using the context menus
 - Allow specific hiding of features, links and chromosomes
*** 0.2.1 <2015-07-16 Do>
 - Added MIT License
 - Included License information for third party libraries
*** 0.2.0 <2015-07-16 Do>
 - allow for input as tsv and bed files as alternative to fasta files
 - demo data added - seven chloroplast genomes
 - documentation added
 - test cases added
 - added interactive JavaScript output
 - renamed to Alignment Toolbox and Visualization (AliTV)
*** 0.1.0 <2015-01-31 Sa>
 - First release of the wgaPipeline code.
 - Automated whole genome alignment and circos visualization from two fasta files.
