#+AUTHOR: Markus Ankenbrand, Frank FÃ¶rster
#+TITLE: Functional specifications for the Alignment Toolbox and Visualization (AliTV)
#+DESCRIPTION: This file contains the functional specification of the AliTV
#+EMAIL: markus.ankenbrand@uni-wuerzburg.de, frank.foerster@biozentrum.uni-wuerzburg.de

#+LATEX_CLASS: scrartcl
#+LATEX_CLASS_OPTIONS: [a4paper,english]
#+LATEX_HEADER: \usepackage[T1]{fontenc}
#+LATEX_HEADER: \usepackage[utf8]{inputenc}

#+TODO: TODO(t!) INPG(i@/!) TEST(n@/!) TESTFAIL(f@/!) TESTPASS(p@/!) | DONE(d!) REJC(c@)


* Objectives of AliTV
  The software Alignment Toolbox and Visualization (AliTV) should be
  used to interactively generate 'publication ready' figures starting
  from biological data sets. Moreover, its capability of changing the
  figure output on the fly extends the functionality of AliTV to an
  analysis tool.
** Must-have criteria
   - Reimplementation of existing functionality including documentation and testing
     - Calculate coordinates for links (linear layout)
     - Draw links (linear layout)
     - Color karyos (linear layout)
     - Add ticks to the karyos (linear layout)
     - Color links based on identity (linear layout)
     - Fade in/out links on mouse over event (linear layout)
     - Reverse karyos and corresponding links by clicking on the karyos
   - Visualization of annotated features on the karyos, e.g. genes or
     inverted repeat regions
   - Visualization of a provided phylogenetic tree next to the alignments
   - Add text labels to the genomes, chromosomes (karyos), and features
   - Implement APIs to the AliTV class to allow remote configuration of
     - graphic parameters of AliTV objects
     - filter parameters of AliTV objects
** Nice-to-have criteria
** (Abgrenzungskriterien)
* Usage
** Application field
** Target audience
** System requirements
* Environment
** Software
** Hardware
** (Orgware)
* Functionality
* Data
* User interface
* Quality targets
